### PR TITLE
Dynamic gas data trie loads

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
## Reasoning behind the pull request
- The new feature of dynamic gas cost computation based on trie loads needs to be activated at a certain epoch by all nodes.
  
## Proposed changes
- Add a new flag `DynamicGasCostForDataTrieStorageLoadEnableEpoch` that will enable at the specified epoch the computation of gas cost for data trie loads based on the depth of the trie.

## Testing procedure
- Create a smart contract with multiple leaves in the dataTrie. When a leaf is accessed, check that the gas cost is computed dynamically based on the depth of the leaf.

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/ElrondNetwork/elrond-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: cmd/node/config/gasSchedules/gasScheduleV7.toml [hunk 1]
```diff
@@ -692,3 +692,14 @@
     MaxBuiltInCallsPerTx = 100
     MaxNumberOfTransfersPerTx = 100
     MaxNumberOfTrieReadsPerTx = 100
+
+# A, B and C are the coefficients for a quadratic func. Separate variables are used for the
+# sign of each coefficient, 0 meaning + and 1 meaning -
+# The current values for the coefficients were computed based on benchmarking
+[DynamicStorageLoad]
+    A = 687
+    SignOfA = 0
```
Comment (BeniaminDrasovean) [inline cmd/node/config/gasSchedules/gasScheduleV7.toml:701]: Looking for better naming...

File: cmd/node/config/enableEpochs.toml [hunk 2]
```diff
@@ -236,6 +236,9 @@
     # WipeSingleNFTLiquidityDecreaseEnableEpoch represents the epoch when the system account liquidity is decreased for wipeSingleNFT as well
     WipeSingleNFTLiquidityDecreaseEnableEpoch = 5
 
+    # DynamicGasCostForDataTrieStorageLoadEnableEpoch represents the epoch when dynamic gas cost for data trie storage load will be enabled
+    DynamicGasCostForDataTrieStorageLoadEnableEpoch = 6
```
Comment (iulianpascalau) [inline cmd/node/config/enableEpochs.toml:240]: switch the value to 3? (what will come from rc/v1.4.0 will have all the other values set to 2)

File: cmd/node/config/gasSchedules/gasScheduleV1.toml [hunk 3]
```diff
@@ -691,3 +691,16 @@
     MaxBuiltInCallsPerTx = 100
     MaxNumberOfTransfersPerTx = 100
     MaxNumberOfTrieReadsPerTx = 100
+
+# Quadratic, Linear and Constant are the coefficients for a quadratic func. Separate variables are used for the
+# sign of each coefficient, 0 meaning + and 1 meaning -
```
Comment (iulianpascalau) [inline cmd/node/config/gasSchedules/gasScheduleV1.toml:696]: `0 meaning positive and 1 meaning negative`?
as it reads better. Valid on all the other gasmaps

File: config/epochConfig.go [hunk 4]
```diff
@@ -91,6 +91,7 @@ type EnableEpochs struct {
 	RefactorPeersMiniBlocksEnableEpoch                uint32
 	MaxBlockchainHookCountersEnableEpoch              uint32
 	WipeSingleNFTLiquidityDecreaseEnableEpoch         uint32
+	DynamicGasCostForDataTrieStorageLoadEnableEpoch   uint32
```
Comment (iulianpascalau) [inline config/epochConfig.go:94]: also adapt tests in the tomlConfig_test.go ?

File: integrationTests/vm/txsFee/relayedScCalls_test.go [hunk 5]
```diff
@@ -11,14 +11,17 @@ import (
 	"testing"
 
 	"github.com/multiversx/mx-chain-go/config"
+	"github.com/multiversx/mx-chain-go/integrationTests"
```
Comment (iulianpascalau) [inline integrationTests/vm/txsFee/relayedScCalls_test.go:14]: please add a semiintegration test here in which you test the transaction fee is computed correctly if the DynamicGasCostForDataTrieStorageLoad is activated. Suggestion to have 2 tests: one when using the SaveKeyValue builtin function and one using a contract that can add data (maybe using an internal array or something).

File: cmd/node/config/enableEpochs.toml (hunk 2 above)
Comment (BeniaminDrasovean) [reply cmd/node/config/enableEpochs.toml:243]: Done.

File: config/epochConfig.go (hunk 4 above)
Comment (BeniaminDrasovean) [reply config/epochConfig.go:95]: Done.

File: integrationTests/vm/txsFee/relayedScCalls_test.go (hunk 5 above)
Comment (BeniaminDrasovean) [reply integrationTests/vm/txsFee/relayedScCalls_test.go:14]: Added.

File: integrationTests/vm/txsFee/dynamicGasCost_test.go [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+	for i, key := range keys {
+		_, depth, err := tr.Get(key)
+		require.Nil(t, err)
+		trieLevels[i] = depth
+	}
+
+	return trieLevels
+}
+
+func deployContract(t *testing.T, testContext *vm.VMTestContext, pathToContract string) []byte {
```
Comment (iulianpascalau) [inline integrationTests/vm/txsFee/dynamicGasCost_test.go:153]: I think we have this function already defined in the "utils" package

File: go.sum [hunk 7]
```diff
@@ -611,14 +611,14 @@ github.com/multiversx/mx-chain-storage-go v1.0.7 h1:UqLo/OLTD3IHiE/TB/SEdNRV1GG2
 github.com/multiversx/mx-chain-storage-go v1.0.7/go.mod h1:gtKoV32Cg2Uy8deHzF8Ud0qAl0zv92FvWgPSYIP0Zmg=
 github.com/multiversx/mx-chain-vm-common-go v1.3.34/go.mod h1:sZ2COLCxvf2GxAAJHGmGqWybObLtFuk2tZUyGqnMXE8=
 github.com/multiversx/mx-chain-vm-common-go v1.3.36/go.mod h1:sZ2COLCxvf2GxAAJHGmGqWybObLtFuk2tZUyGqnMXE8=
-github.com/multiversx/mx-chain-vm-common-go v1.3.37 h1:KeK6JCjeNUOHC5Z12/CTQIa8Z1at0dnnL9hY1LNrHS8=
-github.com/multiversx/mx-chain-vm-common-go v1.3.37/go.mod h1:sZ2COLCxvf2GxAAJHGmGqWybObLtFuk2tZUyGqnMXE8=
+github.com/multiversx/mx-chain-vm-common-go v1.3.38-0.20230227140728-5bd58c1e7990 h1:AYpgFg+jhi90nuxK8sRFyFPovSvQgbF4wdo025SrsFE=
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply go.sum:614]: Will add a proper release right before merge

File: integrationTests/testInitializer.go [hunk 8]
```diff
@@ -2538,3 +2539,38 @@ func SaveDelegationContractsList(nodes []*TestProcessorNode) {
 		_, _ = n.AccntState.Commit()
 	}
 }
+
+var charsPool = []string{"0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "A", "B", "C", "D", "E"}
+
+func GenerateTrieKeysForMaxLevel(numTrieLevels int, numChildrenPerBranch int) [][]byte {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply integrationTests/testInitializer.go:2545]: Added.

File: integrationTests/vm/txsFee/dynamicGasCost_test.go [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+	sndNonce += uint64(len(keys))
+
+	dataTrie := getAccountDataTrie(t, testContext, scAddress)
+	trieKeysDepth := getTrieDepthForKeys(t, dataTrie, keys)
+
+	apiCallsCost := 3
+	loadValCost := 32
+	wasmOpsCost := 14
+
+	newContractCode := wasm.GetSCCode("../wasm/testdata/trieStoreAndLoad/storage.wasm")
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply integrationTests/vm/txsFee/dynamicGasCost_test.go:61]: Done.

File: integrationTests/vm/txsFee/dynamicGasCost_test.go [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+	latestGasSchedule := gasScheduleNotifier.LatestGasSchedule()
+	AoTPrepare := latestGasSchedule[common.BaseOperationCost]["AoTPreparePerByte"] * uint64(len(newContractCode)) / 2
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply integrationTests/vm/txsFee/dynamicGasCost_test.go:63]: Done

File: integrationTests/vm/txsFee/dynamicGasCost_test.go [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+	nonce uint64,
+	maxTrieLevel int,
+	sndAddr []byte,
+	scAddress []byte,
+	gasPrice uint64,
+	gasLimit uint64,
+) [][]byte {
+
+	keys := integrationTests.GenerateTrieKeysForMaxLevel(maxTrieLevel, 2)
+	for _, key := range keys {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply integrationTests/vm/txsFee/dynamicGasCost_test.go:91]: Done.

File: integrationTests/vm/txsFee/dynamicGasCost_test.go (hunk 6 above)
Comment (BeniaminDrasovean) [reply integrationTests/vm/txsFee/dynamicGasCost_test.go:153]: Indeed. Used the func from utils.

File: config/tomlConfig_test.go [hunk 12]
```diff
@@ -779,6 +783,7 @@ func TestEnableEpochConfig(t *testing.T) {
 			AlwaysSaveTokenMetaDataEnableEpoch:          61,
 			RuntimeCodeSizeFixEnableEpoch:               62,
 			RuntimeMemStoreLimitEnableEpoch:             63,
+			DynamicGasCostForDataTrieStorageLoadEnableEpoch: 64,
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply config/tomlConfig_test.go]: Done.

**Review by gabi-vuls — APPROVED** [review]
Normal allin test: master-0a0bcee937 -> dynamic-gas-data-trie-load-2a6908a5d3

--- Specific errors ---

block hash does not match 8743
wrong nonce in block 3105
miniblocks does not match 0
num miniblocks does not match 0
miniblock hash does not match 0
block bodies does not match 0
receipts hash missmatch 0

/------/


--- Statistics ---

Nr. of all ERRORS:  3
Nr. of all WARNS:  145
Nr. of new ERRORS:  0
Nr. of new WARNS:  5
Nr. of PANICS:  0

/------/

**Review by iulianpascalau — COMMENTED** [review]
just the proper release

**Review by iulianpascalau — APPROVED** [review]

**Review by ssd04 — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
Normal allin test: master-0a0bcee937 -> dynamic-gas-data-trie-load-12a225a566

--- Specific errors ---

block hash does not match 1080
wrong nonce in block 467
miniblocks does not match 0
num miniblocks does not match 0
miniblock hash does not match 0
block bodies does not match 0
receipts hash missmatch 0

/------/


--- Statistics ---

Nr. of all ERRORS:  0
Nr. of all WARNS:  133
Nr. of new ERRORS:  0
Nr. of new WARNS:  0
Nr. of PANICS:  0

/------/

--- ERRORS ---

/------/

--- WARNINGS ---

/------/

**Pull request merged**